### PR TITLE
Highlighting without geniplate

### PR DESCRIPTION
**Pull request by andreasabel**
[ refactor for #2038 ] highlighting: nameInfo/theRest w/o geniplate

This refactoring implements the extraction of highlighting information from abstract syntax by a manual traversal.  Previously, this traversal was split into 13 individual traversals automatically generated via geniplate.

The goal is to get rid of geniplate which has been abandoned by the original author, putting the burden of maintenance on us.  Getting rid of geniplate will also dispose of the use of TemplateHaskell for Abstract.hs, which noticeably slows compilation.

This patch does not reach the goal completely, as the NameKinds map is still constructed using geniplate.  The NameKind information could instead be maintained in the ScopeInfo.

A more radical refactoring would get entirely rid of a separate highlighting pass over abstract syntax and instead record information about highlighting in the TCState, incremently adding information whenever the kind of a name is resolved in the scope checker.  We already do a similar thing for the disambiguation of overloaded symbols in the type checker.

### CODE REVIEW

File: src/full/Agda/Interaction/Highlighting/FromAbstract.hs [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+      A.DataDef _di x _uc pars cs            -> hl x <> hl pars <> hl cs
+      A.RecSig _di x tel e                   -> hl x <> hl tel <> hl e
+      A.RecDef _di x _uc _ind _eta y bs e ds -> hl x <> hl y <> hl bs <> hl e <> hl ds
+      A.PatternSynDef x xs p                 -> hl x <> hl xs <> hl p
+      A.UnquoteDecl _mi _di xs e             -> hl xs <> hl e
+      A.UnquoteDef _di xs e                  -> hl xs <> hl e
+      A.ScopedDecl s ds                      -> hl ds
+      A.Pragma _r pragma                     -> hl pragma
+    where
+    hl      a = hilite a
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/full/Agda/Interaction/Highlighting/FromAbstract.hs:214]: I seem to remember that eta-reduction fails type checking in this case.

**Comment by asr** [comment]
I suggest to add the text "complete tests" to the commit message for testing this PR with all the supported versions of GHC.

**Comment by nad** [comment]
[Elsewhere](https://github.com/agda/agda/issues/2038#issuecomment-571784173) I wrote the following:

> For anyone who wants to try this:
> * Test that the changes do not affect the HTML that is generated for the standard library. (I doubt that we have test cases for the highlighting of every kind of Agda syntax.)
> * Benchmark Agda using `make library-test`.

@andreasabel, did you do this?

**Comment by andreasabel** [comment]
@asr:
> I suggest to add the text "complete tests" to the commit message for testing this PR with all the supported versions of GHC.

Ah, I did not know about this flag.  Aren't all GHC versions tested by the github actions?

**Comment by andreasabel** [comment]
@nad wrote:
> > * Test that the changes do not affect the HTML that is generated for the standard library. (I doubt that we have test cases for the highlighting of every kind of Agda syntax.)

The test suite is basically our specification and defines the boundaries in which the Agda organism evolves.  Features we care about should be secured by the test suite.  Having a HTML test for the whole library seems infeasible, because the library changes all the time.

I have been working with the LaTeXAndHTML tests so far which catch a lot of issues (but are not complete, see #4580).

If there is a special highlighting feature you care about that is not covered in LaTeXAndHTML, please add it to this suite.

> > * Benchmark Agda using `make library-test`.

This looks neutral (no dramatic change in the time spent on highlighting).

Should be looked at again when this PR is complete.  I am currently working to get the `NameKind` from what the scope checker produces, to get rid of the extra traversal of declarations in the highlighter (remaining use of geniplate).

**Comment by asr** [comment]
> Aren't all GHC versions tested by the github actions?

Only compilation in all the supported GHC versions is tested by the GitHub workflows (except GHC 8.10.1 but it is due to one knowed issue described in #4242).

**Comment by andreasabel** [comment]
@asr:  You got me confused:
> I suggest to add the text "complete tests" to the commit message for testing this PR with all the supported versions of GHC.

> Only compilation in all the supported GHC versions is tested by the GitHub workflows

Let me ask it differently: Which additional tests are run if I include "complete tests"?  Which additional GHC versions?

**Comment by asr** [comment]
By including `complete tests` you run on Travis the following tests with all the supported GHC versions:
https://github.com/agda/agda/blob/9573514d85cc600bdb4bcad1d16fcb96283f029e/.travis.yml#L198-L227

**Comment by andreasabel** [comment]
@asr:  Thanks for the explanation.  I now put "complete tests" in the commit.  I still do not see why it would be especially relevant for this PR.  I would understand if I changed the GHC backend; in this case, the compiler test should run on all GHC versions.  For highlighting, I don't see how different GHC versions should behave differently (except for GHC bugs, but this is nothing particular to the highlighting feature).

**Comment by andreasabel** [comment]
The implementation of this PR is more conservative than I intended.  I spent two days to get highlighting entirely out of the results of the scope checker (except for ambiguous constructors/projections).  I used `inverseScopeLookup` to compute the `NameKind` of a `A.QName`.  Everything worked, but this was a big performance regression (12% of std-lib time spend on inverseScopeLookup).  The inverse scope maps are apparently very inefficiently computed, and recomputed on every scope update.  Usually, they are only used in the printer, thus, the inefficiency does not matter too much.  They are not up to serious use yet, unfortunately.

**Comment by andreasabel** [comment]
@asr: The effect of "complete tests" was that the compiler test timed out on 8.0.2, as already observed so often before (#4018, #4177). Not sure why I want to go back to this.

**Comment by ajrouvoet** [comment]
@andreasabel @jespercockx the performance of reverseScopeLookup was also a concern for the implementation of `--no-qualified-instances` IIRC, which uses it to decide if an instance needs to be considered a candidate.

**Comment by nad** [comment]
> The test suite is basically our specification and defines the boundaries in which the Agda organism evolves. Features we care about should be secured by the test suite. Having a HTML test for the whole library seems infeasible, because the library changes all the time.

One can run `agda --html` twice for the standard library, and check that the generated HTML files are identical. I'm not suggesting that we should do this all the time, but it might have made sense for this large refactoring.

**Comment by nad** [comment]
I ran `agda --html` using two recent versions of Agda and found quite a few differences. I did not look at all of them. The following might be related to this pull request:
```agda
record R : Set₁ where
  field
    A : Set

postulate
  r : R

open R r

r₁ : R
r₁ = record { A = A }

r₂ : R
r₂ = record { A = R.A r }
```
Previously the final occurrence of `A` in `r₁` was highlighted as a function, now it is highlighted as a field.

**Comment by ajrouvoet** [comment]
I've noticed before that (even unqualified) fields were highlighted as functions. So some differences might be an improvement rather than a regression.

**Comment by gallais** [comment]
Cf. #3156 too

**Comment by asr** [comment]
@andreasabel, I'm concern about we are running all the tests for all the supported versions of GHC by default _only_ in the `releases` branches. For this PR, I wanted to see green in Travis for all GHC versions.

**Pull request merged**